auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FailoverSecurityRealm.java

```diff
@@ -0,0 +1,272 @@
+    public RealmIdentity getRealmIdentity(final Principal principal) throws RealmUnavailableException {
+        try {
+            return createFailoverIdentity(delegateRealm.getRealmIdentity(principal), principal);
+        } catch (RealmUnavailableException e) {
+            log.realmFailover(e);
+            if (failoverCallback != null) {
+                failoverCallback.accept(e);
+            }
+            return failoverRealm.getRealmIdentity(principal);
+        }
+    }
+
+    @Override
+    public SupportLevel getCredentialAcquireSupport(final Class<? extends Credential> credentialType, final String algorithmName, final AlgorithmParameterSpec parameterSpec) throws RealmUnavailableException {
+        return delegateRealm.getCredentialAcquireSupport(credentialType, algorithmName, parameterSpec);
+    }
+
+    @Override
+    public SupportLevel getEvidenceVerifySupport(final Class<? extends Evidence> evidenceType, final String algorithmName) throws RealmUnavailableException {
+        return delegateRealm.getEvidenceVerifySupport(evidenceType, algorithmName);
+    }
+
+    protected RealmIdentity createFailoverIdentity(final RealmIdentity identity, final Evidence evidence) {
+        return new FailoverRealmIdentity(identity) {
+            @Override
+            protected RealmIdentity getFailoverIdentity() throws RealmUnavailableException {
+                return failoverRealm.getRealmIdentity(evidence);
+            }
+        };
+    }
+
+    protected RealmIdentity createFailoverIdentity(final RealmIdentity identity, final Principal principal) {
+        return new FailoverRealmIdentity(identity) {
+            @Override
+            protected RealmIdentity getFailoverIdentity() throws RealmUnavailableException {
+                return failoverRealm.getRealmIdentity(principal);
+            }
+        };
+    }
+
+    protected abstract class FailoverRealmIdentity implements RealmIdentity {
+        protected RealmIdentity delegate;
+        protected boolean failed = false;
+
+        public FailoverRealmIdentity(final RealmIdentity identity) {
+            this.delegate = identity;
+        }
+
+        protected abstract RealmIdentity getFailoverIdentity() throws RealmUnavailableException;
+
+        @Override
+        public SupportLevel getCredentialAcquireSupport(Class<? extends Credential> credentialType, String algorithmName, final AlgorithmParameterSpec parameterSpec) throws RealmUnavailableException {
+            return delegate.getCredentialAcquireSupport(credentialType, algorithmName, parameterSpec);
+        }
+
+        @Override
+        public <C extends Credential> C getCredential(Class<C> credentialType) throws RealmUnavailableException {
+            try {
+                return delegate.getCredential(credentialType);
+            } catch (RealmUnavailableException rue) {
+                return failover(rue).getCredential(credentialType);
+            } finally {
+                disableFailover();
```

Comment:
I think it would be good to have some comments that explain why `disableFailover()` is called in some methods but not in others.

Reply:
Generally speaking this is to prevent failover after the identity sends or recieves any authentication data, so it cannot failover in the middle of authentication. I've added a single line comment to the method as an explanation.

Reply:
Thanks, Martin.